### ISSUE TITLE
Consider the minimum possible SemVer bump when selecting lints for "no version change" runs

### ISSUE THREAD

**Issue by jonhoo** [obi1kenobi/cargo-semver-checks#1207]
### Which lint or lints are the issue

`inherent_method_must_use_added`

### Known issues that might be causing this

- [ ] Is the flagged item defined in another crate, or a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))

### Steps to reproduce the bug with the above code

Trigger any `required_update: Minor` lint in a crate versioned at `0.1.0`.

### Actual Behaviour

The `required_update: Minor` check fails with:
```
semver requires new minor version: 0 major and 1 minor checks failed
```

### Expected Behaviour

No `required_update: Minor` check should fail.

### Verbose Lint Output

```

--- failure inherent_method_must_use_added: inherent method #[must_use] added ---

Description:
An inherent method is now #[must_use]. Downstream crates that did not use its return value will get a compiler lint.
        ref: https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute
       impl: https://github.com/obi1kenobi/cargo-semver-checks/tree/v0.40.0/src/lints/inherent_method_must_use_added.ron

Failed in:
  method fantoccini::client::RawRequestBuilder::map_request in /home/jon/dev/minor/fantoccini/src/client.rs:320
```

### Generated System Information

System information:
-------------------
#### Software version

cargo-semver-checks 0.40.0

#### Operating system

- OS: Linux (Arch Linux rolling)
- Kernel: 6.13.5-arch1-1

#### Command-line

```bash
/home/jon/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.86.0-beta.6 (adf9b6ad1 2025-02-28)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

### Additional Context

MAINTAINER EDIT: See [explanation here](https://github.com/obi1kenobi/cargo-semver-checks/issues/1207#issuecomment-2745331161) for full details; the below text is a summary. I'm leaving the original text of this issue below.

When `cargo-semver-checks` is run on a `current` and `baseline` package that share the same version, it doesn't know what SemVer change the maintainer _intends_ to make in the future. Clearly they cannot publish the changes under the existing, already published version!

The current behavior in this case is that we run _all_ lints. This issue makes a convincing case that we should consider the _minimum possible_ SemVer change for the next release, and only run lints appropriate for that. For example, if `baseline` is v0.1.0, any subsequent version is either minor or major, not patch, meaning we shouldn't run any lints that demand a minor version since they will always be satisfied here.

ORIGINAL:

I just had the `inherent_method_must_use_added` lint trigger in a crate currently versioned at `0.12.1`. Putting aside the fact that I think lint-triggering changes should not be considered to warrant a minor release, I think specifically in pre-1.0 releases, we should probably consider every release that bumps the patch number (i.e., 0.X.Y -> 0.X.Y+1) to be a _minor_ release not a patch release for the purpose of semver checking. After all, the minor number is considered the major version number at pre-1.0, so wouldn't that also "shift" the semantics of the second number?

**Comment by obi1kenobi** [comment]
Thanks for opening the issue. Just to help pin down why you're seeing this, can you find the output line that says `Checking fantoccini 0.12.1 -> ...` and let me know what it says?

It will likely either be:
```
Checking fantoccini 0.12.1 -> 0.12.1 (no change)
```
or
```
Checking fantoccini 0.12.1 -> 0.12.2 (minor change)
```

This will help us pin down what version change `cargo-semver-checks` *thinks* is happening, so that we know if the bug is an implementation error or a correct implementation of a suboptimal workflow.

**Comment by obi1kenobi** [comment]
Aside: our docs have a section on [downgrading `#[must_use]` lints to just a warning or disabling them entirely](https://github.com/obi1kenobi/cargo-semver-checks?tab=readme-ov-file#common-configurations-make-must_use-lints-warn-only). It's not an unreasonable nor uncommon configuration to run `cargo-semver-checks` in — the defaults err on the side of being more conservative for sure.

**Comment by jonhoo** [comment]
```
    Checking fantoccini v0.21.4 -> v0.21.4 (no change)
```
This was in an MR that just made the relevant change and did not modify `Cargo.toml` at all (which is generally what I expect people to do in MRs).

On the aside: heh, yeah, I have _opinions_ on
> The rationale is that such an addition risks introducing new lints in downstream projects, and many projects consider lints as errors and may be broken as a result. (Whether that setting is good practice for widespread use or not is outside the scope of cargo-semver-checks.)
though I very much see how you arrived at the current policy. In a way, my take here is that picking for it to be a semver-checks error by default _is_ taking a stand on whether it's good practice. Just as not making it an error by default would be.

**Comment by obi1kenobi** [comment]
> ```
>     Checking fantoccini v0.21.4 -> v0.21.4 (no change)
> ```
> 
> This was in an MR that just made the relevant change and did not modify `Cargo.toml` at all (which is generally what I expect people to do in MRs).

Ah yes, that's what I suspected.

The current behavior is this: when the `baseline` and `current` packages have the same version, we enable all lints. This is _regardless_ of version numbers.

If the version number had changed from `0.21.4 -> 0.21.5` that would have been detected as a minor bump (not patch) and the SemVer-minor lints would have been skipped.

The intent behind running all lints when the versions are the same is that we don't know what version the maintainer _plans_ to release that change in, since it clearly can't be the current one.

I think I could be convinced that the code should also take into account the minimum version bump they possibly could make (e.g. there's no possible patch bump from `0.21.4`, only minor or major), and skip lints accordingly. It sounds like that's what you were suggesting as well?

> On the aside: heh, yeah, I have _opinions_ on
> 
> > The rationale is that such an addition risks introducing new lints in downstream projects, and many projects consider lints as errors and may be broken as a result. (Whether that setting is good practice for widespread use or not is outside the scope of cargo-semver-checks.)
> 
> though I very much see how you arrived at the current policy. In a way, my take here is that picking for it to be a semver-checks error by default _is_ taking a stand on whether it's good practice. Just as not making it an error by default would be.

The current policy is a product of:
- Lints that are disabled by default are lints that ~nobody will find. (Have you noticed any of the allow-by-default lints we ship? I'd bet not! We have no discovery mechanism for them.) On the flip side, we have a pretty easy opt-out if someone doesn't like those lints.
- A poll I ran on Mastodon / bsky a year+ ago that asked Rustaceans if triggering lints downstream should be major, minor, or patch by default in cargo-semver-checks. IIRC minor won decisively, though the poll had a small number of votes.

I know anything we choose _is_ taking a stand 😅 I could be convinced to make the lints warn-by-default, if, say, a prominent Rustacean wrote or recorded a convincing argument about it 👀 Perhaps also a good topic to discuss in r/rust or a similar venue where lots of cargo-semver-checks users would have an opportunity to express their desired defaults and reasons? I think allow-by-default wouldn't be the right choice right now, though.

**Comment by jonhoo** [comment]
> I think I could be convinced that the code should also take into account the minimum version bump they possibly could make (e.g. there's no possible patch bump from 0.21.4, only minor or major), and skip lints accordingly. It sounds like that's what you were suggesting as well?

Yep, that sounds exactly right!

> The current policy is a product of:

To be clear, I wasn't intending to criticize — you had to pick _some_ default, and there _are_ arguments on either side. I'm just so firmly on the side of "warnings should not be treated as errors" that for anything that's counter to that I can't help but remark on it.

> Lints that are disabled by default are lints that ~nobody will find.

100% acknowledged. However, my argument here is that this should not be treated as an error _by anyone ever_, so I'm fine with it disappearing into a dark spot where no one will find it ever again 😅 

> I could be convinced to make the lints warn-by-default, if, say, a prominent Rustacean wrote or recorded a convincing argument about it

Oh, you mean like [giving a keynote at Rust Nation making that argument](https://youtu.be/r35cBkPRNMI?t=249)? :p

Slightly more seriously though: I agree broader consensus in the community here would be nice, but sadly I don't think we will get it any time soon. And the reason is simple: at that moment, basically no CI tooling I know of handles warnings in a useful way. Your options are to error out or miss the warnings completely. And when faced with that choice, it's _real_ tempting to say "well, then, warnings have to be errors I guess". Whereas I'd argue it's a false choice, and that the answer is to build better tooling for surfacing warnings. GitHub annotations tried to be that, but they were also completely ignored, probably because they never make it into the main thread of conversation and so easily get lost. [ReviewDog](https://github.com/reviewdog/reviewdog) have another option that seems quite good, but I haven't seen it adopted anywhere, probably because it requires you actually set it up. But what you end up with then is _far_ better than failing CI, because it has all the upsides and none of the downsides.

**Comment by obi1kenobi** [comment]
> Yep, that sounds exactly right!

Neat! I think it probably would be a contributor-friendly change, so let me retitle the issue to make it clearer what contributors should do.

> To be clear, I wasn't intending to criticize

No worries, no criticism perceived :)

> Oh, you mean like [giving a keynote at Rust Nation making that argument](https://youtu.be/r35cBkPRNMI?t=249)?

I was hoping you'd say something like that! 😂

100% agreed on the unfortunate state of warnings in Rust. GitHub Annotations are also comically bad in some cases: for example, GitHub collects warnings from our test crates at paths like `test_crates/some_crate/src/lib.rs` and then displays annotations on `src/lib.rs` of `cargo-semver-checks` itself instead of the test crate 🤦

There's a student working on a better GitHub Action for `cargo-semver-checks` where we hope to do a better job of surfacing warnings, but it's still a ways out. I hope we can be (a small part of) the change we wish to see in the world!

**Comment by obi1kenobi** [comment]
> Do you think this is reasonable, when base version is equal to current version:
> 
> * For 0.0.z: Minimum next change must be major
> * For 0.y.z (y > 0): Minimum next change is minor
> * For stable versions x.y.z (x ≥ 1): Minimum next change is patch

Yes, that's also what I was thinking :+1:

> do you have any idea on when we will want to run all lints? (like when `BuildMetadata` is different)

I think `BuildMetadata` shouldn't influence our lint selection. In the long run, I expect we'll have lint groups and the groups will dictate what gets run when. There may be lints that always run regardless of version (#1124), and some lints may run on only major/minor versions etc. But I don't think we have to design all that today — we can solve this issue with a more focused change and just leave room for future growth for the rest.

**Comment by Frank-III** [comment]
>Yes, that's also what I was thinking 👍
>I think BuildMetadata shouldn't influence our lint selection. 

would work on it then!

>I expect we'll have lint groups and the groups will dictate what gets run when. There may be lints that always run regardless of version (https://github.com/obi1kenobi/cargo-semver-checks/issues/1124), and some lints may run on only major/minor versions etc. But I don't think we have to design all that today — we can solve this issue with a more focused change and just leave room for future growth for the rest.

gotcha, I was thinking if I could make implementing lint group my GSoC proposal, would love to be involved if there are more discussions around it!

**Comment by obi1kenobi** [comment]
If you're interested in working on lint groups as part of GSoC, this issue can definitely fall under that umbrella as well. It would be particularly nice to have a single person as the owner of the entire lint selections user experience end-to-end!
